backend/utils/pubsub.js

```diff
@@ -0,0 +1,261 @@
+/* globals logger */
+'use strict'
+
+import {
+  NOTIFICATION_TYPE,
+  REQUEST_TYPE,
+  RESPONSE_TYPE,
+  createClient,
+  createMessage,
+  messageParser
+} from '~/shared/pubsub.js'
+
+import type { Message, SubMessage, UnsubMessage } from '~/shared/pubsub.js'
+import type { JSONType } from '~/shared/types.js'
+
+const { bold } = require('chalk')
+const WebSocket = require('ws')
+
+// ====== API ====== //
+
+const { PUB, SUB, UNSUB } = NOTIFICATION_TYPE
+const { ERROR, SUCCESS } = RESPONSE_TYPE
+
+// Re-export some useful things from the shared module.
+export { createClient, createMessage, NOTIFICATION_TYPE, REQUEST_TYPE, RESPONSE_TYPE }
+
+export function createNotification (
+  type: string,
+  data: {| contractID: string, socketID: number |}
+): string {
+  return JSON.stringify({ type, ...data })
+}
+
+export function createResponse (type: string, data: JSONType): string {
+  return JSON.stringify({ type, data })
+}
+
+/**
+ * Creates a pubsub server instance.
+ *
+ * @param {(http.Server|https.Server)} server - A Node.js HTTP/S server to attach to.
+ * @param {Object?} options
+ * {object?} clientHandlers - Custom handlers for socket events.
+ * {object?} messageHandlers - Custom handlers for different message types.
+ * {object?} serverHandlers - Custom handlers for server events.
+ *
+ * {number} backlog - The maximum length of the queue of pending connections.
+ * {boolean} clientTracking - Specifies whether or not to track clients.
+ * {Function} handleProtocols - A function which can be used to handle the WebSocket subprotocols.
+ * {number} maxPayload - The maximum allowed message size in bytes.
+ * {string} path - Accept only connections matching this path.
+ * {(boolean|object)} perMessageDeflate - Enables/disables per-message deflate.
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = { ...options.clientHandlers }
+  server.customServerHandlers = { ...options.serverHandlers }
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
```

Comment:
Yes, at least in theory, because `Map` objects currently do not appear to be used anywhere in the codebase, despite likely more than one appropriate use case. Perhaps because of their API which might feel heavier than object property access.
So I've used an object for consistency. It's fine to be used as a map since all keys are going to be strings and a null prototype object doen't initially have any property.

Reply:
Okay, that makes sense. Probably something to evaluate/fix as a separate point in the future, then.